cirq/ops/fsim_gate.py

```diff
@@ -164,3 +165,268 @@ def __repr__(self) -> str:
 
     def _json_dict_(self) -> Dict[str, Any]:
         return protocols.obj_to_dict_helper(self, ['theta', 'phi'])
+
+
+@value.value_equality(approximate=True)
+class PhasedFSimGate(gate_features.TwoQubitGate,
+                     gate_features.InterchangeableQubitsGate):
+    """General excitation-preserving two-qubit gate.
+
+    The unitary matrix of this gate is:
+
+        [[1,           0,             0,           0],
+         [0,    f1 cos(θ), -i f2 sin(θ),           0],
+         [0, -i f3 sin(θ),    f4 cos(θ),           0],
+         [0,           0,             0, f5 exp(-iφ)]]
+
+    where the phase factors are
+
+        f1 = exp(i∆p + i∆m)
+        f2 = exp(i∆p - i∆mod)
+        f3 = exp(i∆p + i∆mod)
+        f4 = exp(i∆p - i∆m)
+        f5 = exp(2i∆p)
+
+    and the correspondence between the symbols above and the parameters
+    to __init__ is as follows:
+
+        θ:    theta,
+        φ:    phi,
+        ∆p:   delta_plus,
+        ∆m:   delta_minus,
+        ∆mod: delta_minus_off_diagonal.
+
+    See also eq (43) in https://arxiv.org/abs/1910.11333.
+
+    There is another useful parametrization of PhasedFSimGate based on
+    the fact that the gate is equivalent to the following circuit:
+
+        0: ───Rz(b0)───FSim(theta, phi)───Rz(a0)───
+                       │
+        1: ───Rz(b1)───FSim(theta, phi)───Rz(a1)───
+
+    where b0 and b1 are phase angles to be applied before the core FSimGate,
+    a0 and a1 are phase angles to be applied after FSimGate and theta and
+    phi specify the core FSimGate. Use the static factory function
+    PhasedFSimGate.from_phase_angles_and_fsim to instantiate the gate
```

Comment:
Internally we also have dataclasses to store just the parameters in these different representations with functions to canonicalize them (since the various phase parameters are periodic) and to convert between them. It might be nice to include that code here, though it might be awkward to support symbolic values so would have to try it out. Note that internally we call the canonical parametrization above `FsimPhase` while this alternate parametrization with per-qubit z rotations is called `FsimRz`. If we stick with that nomenclature then the factory function would better be called `PhasedFSimGate.from_fsim_rz` or something like that, and the properties for getting the rotation angles before and after might also best be renamed as `rz_angles_before` and `rz_angles_after`.

Reply:
+1, it would be nice to follow up on this because it will be one of the building blocks of the some calibration features.

Reply:
Renamed the factory method and the two properties as suggested. As discussed on a different PR, I'd like to defer dataclasses (and possibly re-use existing code that @mrwojtek is introducing elsewhere).